File: skills/openclix-init/templates/react-native/domain/CampaignProcessor.ts
import type {
  Campaign,
  TriggerContext,
  CampaignStateSnapshot,
  CampaignStateRecord,
  DecisionTrace,
  QueuedMessage,
  Settings,
  SkipReason,
  Logger,
  DayOfWeek,
  RecurringTriggerConfig,
} from './ClixTypes';
import {
  EventConditionProcessor,
  ScheduleCalculator,
  renderTemplate,
  generateUUID,
} from './CampaignUtils';

export interface CampaignDecision {
  trace: DecisionTrace;
  action: 'trigger' | 'skip';
  queued_message?: QueuedMessage;
  scheduled_for?: string;
}

export interface CampaignProcessorDeps {
  eventConditionProcessor: EventConditionProcessor;
  scheduleCalculator: ScheduleCalculator;
  logger: Logger;
  settings?: Settings;
}

function createTrace(
  campaign_id: string,
  action: string,
  result: 'applied' | 'skipped',
  reason: string,
  skip_reason?: SkipReason,
): DecisionTrace {
  return { campaign_id, action, result, skip_reason, reason };
}

function createSkipDecision(
  campaignId: string,
  reason: string,
  skipReason?: SkipReason,
): CampaignDecision {
  return {
    action: 'skip',
    trace: createTrace(
      campaignId,
      'skip_campaign',
      'skipped',
      reason,
      skipReason,
    ),
  };
}

interface ExecutionResolution {
  execute_at?: string;
  scheduled_for?: string;
  trigger_event_id?: string;
  reason?: string;
  skip_reason?: SkipReason;
}

const DAY_INDEX: Record<DayOfWeek, number> = {
  sunday: 0,
  monday: 1,
  tuesday: 2,
  wednesday: 3,
  thursday: 4,
  friday: 5,
  saturday: 6,
};

function isValidDateValue(value: string | undefined): value is string {
  if (!value) return false;
  return !Number.isNaN(new Date(value).getTime());
}

function startOfWeek(date: Date): Date {
  const copy = new Date(date);
  copy.setHours(0, 0, 0, 0);
  copy.setDate(copy.getDate() - copy.getDay());
  return copy;
}

function withTime(base: Date, hour: number, minute: number): Date {
  const copy = new Date(base);
  copy.setHours(hour, minute, 0, 0);
  return copy;
}

export class CampaignProcessor {
  process(
    campaignId: string,
    campaign: Campaign,
    context: TriggerContext,
    snapshot: CampaignStateSnapshot,
    dependencies: CampaignProcessorDeps,
  ): CampaignDecision {
    const { eventConditionProcessor, scheduleCalculator, logger, settings } = dependencies;
    const now = context.now ?? new Date().toISOString();
    const campaignState = this.getCampaignState(snapshot, campaignId);

    // 1. Check campaign status
    if (campaign.status !== 'running') {
      return createSkipDecision(
        campaignId,
        `Campaign status is '${campaign.status}', not 'running'`,
        'campaign_not_running',
      );
    }

    if (
      (campaign.trigger.type === 'event' && context.trigger !== 'event_tracked') ||
      (campaign.trigger.type !== 'event' && context.trigger === 'event_tracked')
    ) {
      return createSkipDecision(
        campaignId,
        `Trigger type '${campaign.trigger.type}' is not eligible for '${context.trigger}'`,
      );
    }

    if (
      campaign.trigger.type !== 'recurring' &&
      campaignState?.triggered === true
    ) {
      return createSkipDecision(campaignId, 'Campaign already triggered');
    }

    // Global frequency cap: count all triggered campaigns in the rolling window.
    if (settings?.frequency_cap) {
      const { max_count, window_seconds } = settings.frequency_cap;
      const windowStart = new Date(
        new Date(now).getTime() - window_seconds * 1000,
      ).toISOString();
      const recent = (snapshot.trigger_history ?? []).filter(
        (row) => row.triggered_at >= windowStart,
      );
      const countInWindow = recent.length;
      if (countInWindow >= max_count) {
        return createSkipDecision(
          campaignId,
          `Frequency cap exceeded (${countInWindow}/${max_count} within ${window_seconds}s)`,
          'campaign_frequency_cap_exceeded',
        );
      }
    }

    if (
      campaign.trigger.type === 'recurring' &&
      this.hasFuturePendingForCampaign(snapshot, campaignId, now)
    ) {
      return createSkipDecision(
        campaignId,
        'Recurring campaign already has a queued message',
      );
    }

    const resolved = this.resolveExecutionTime(
      campaign,
      context,
      campaignState,
      now,
      eventConditionProcessor,
    );
    if (!resolved.execute_at) {
      return createSkipDecision(
        campaignId,
        resolved.reason ?? 'Campaign trigger conditions were not met',
        resolved.skip_reason,
      );
    }

    if (this.hasPendingForCampaignAt(snapshot, campaignId, resolved.execute_at)) {
      return createSkipDecision(
        campaignId,
        `Duplicate schedule prevented for campaign at ${resolved.execute_at}`,
      );
    }

    const scheduleResult = scheduleCalculator.calculate({
      now,
      execute_at: resolved.execute_at,
      do_not_disturb: settings?.do_not_disturb,
    });

    if (scheduleResult.skipped) {
      return createSkipDecision(
        campaignId,
        'Blocked by do-not-disturb window',
        scheduleResult.skip_reason,
      );
    }

    // 7. Render message content
    const templateVars: Record<string, unknown> = {
      ...(context.event?.properties ?? {}),
    };

    const renderedTitle = renderTemplate(campaign.message.content.title, templateVars);
    const renderedBody = renderTemplate(campaign.message.content.body, templateVars);

    const queuedMessage: QueuedMessage = {
      id: generateUUID(),
      campaign_id: campaignId,
      channel_type: campaign.message.channel_type,
      status: 'scheduled',
      content: {
        title: renderedTitle,
        body: renderedBody,
        image_url: campaign.message.content.image_url,
        landing_url: campaign.message.content.landing_url,
      },
      trigger_event_id: resolved.trigger_event_id,
      execute_at: scheduleResult.execute_at,
      created_at: now,
    };

    logger.debug(
      `[CampaignProcessor] Campaign ${campaignId}: triggered, scheduled for ${scheduleResult.execute_at}`,
    );

    return {
      action: 'trigger',
      trace: createTrace(
        campaignId,
        'trigger_campaign',
        'applied',
        `Campaign triggered, message scheduled for ${scheduleResult.execute_at}`,
      ),
      queued_message: queuedMessage,
      scheduled_for: resolved.scheduled_for ?? scheduleResult.execute_at,
    };
  }

  private resolveExecutionTime(
    campaign: Campaign,
    context: TriggerContext,
    campaignState: CampaignStateRecord | undefined,
    now: string,
    eventConditionProcessor: EventConditionProcessor,
  ): ExecutionResolution {
    if (campaign.trigger.type === 'event') {
      const eventConfig = campaign.trigger.event;
      if (!eventConfig) {
        return {
          reason: "Trigger type 'event' requires trigger.event configuration",
          skip_reason: 'trigger_event_not_matched',
        };
      }
      if (!context.event) {
        return {
          reason: 'Event trigger requires an event in context',
          skip_reason: 'trigger_event_not_matched',
        };
      }
      const matched = eventConditionProcessor.process(
        eventConfig.trigger_event,
        context.event,
      );
      if (!matched) {
        return {
          reason: `Trigger event conditions did not match event '${context.event.name}'`,
          skip_reason: 'trigger_event_not_matched',
        };
      }
      const delaySeconds = eventConfig.delay_seconds ?? 0;
      const executeAt = new Date(new Date(now).getTime() + delaySeconds * 1000).toISOString();
      return {
        execute_at: executeAt,
        scheduled_for: executeAt,
        trigger_event_id: context.event.id,
      };
    }

    if (campaign.trigger.type === 'scheduled') {
      const scheduled = campaign.trigger.scheduled;
      if (!scheduled || !isValidDateValue(scheduled.execute_at)) {
        return { reason: 'Scheduled trigger requires a valid execute_at datetime' };
      }
      const executeAtMs = new Date(scheduled.execute_at).getTime();
      const nowMs = new Date(now).getTime();
      if (executeAtMs <= nowMs) {
        return { reason: `Scheduled execute_at '${scheduled.execute_at}' is already in the past` };
      }
      return {
        execute_at: scheduled.execute_at,
        scheduled_for: scheduled.execute_at,
      };
    }

    const recurring = campaign.trigger.recurring;
    if (!recurring) {
      return { reason: 'Recurring trigger requires recurring configuration' };
    }
    const lastScheduledAt = campaignState?.recurring_last_scheduled_at;
    const recurringAnchorAt = campaignState?.recurring_anchor_at;
    const nextExecuteAt = this.computeNextRecurringExecuteAt(
      recurring,
      now,
      lastScheduledAt,
      recurringAnchorAt,
    );
    if (!nextExecuteAt) {
      return { reason: 'Recurring schedule has no upcoming execution window' };
    }
    return {
      execute_at: nextExecuteAt,
      scheduled_for: nextExecuteAt,
    };
  }

  private computeNextRecurringExecuteAt(
    recurring: RecurringTriggerConfig,
    now: string,
    lastScheduledAt?: string,
    recurringAnchorAt?: string,
  ): string | null {
    const nowDate = new Date(now);
    if (Number.isNaN(nowDate.getTime())) return null;

    const startDate =
      recurring.start_at && isValidDateValue(recurring.start_at)
        ? new Date(recurring.start_at)
        : recurringAnchorAt && isValidDateValue(recurringAnchorAt)
          ? new Date(recurringAnchorAt)
          : this.computeDefaultRecurringAnchorDate(recurring, nowDate);
    if (Number.isNaN(startDate.getTime())) return null;

    const endAtMs = recurring.end_at ? new Date(recurring.end_at).getTime() : Number.POSITIVE_INFINITY;
    if (Number.isNaN(endAtMs)) return null;

    const fromDate =
      lastScheduledAt && isValidDateValue(lastScheduledAt)
        ? new Date(new Date(lastScheduledAt).getTime() + 60_000)
        : nowDate;

    const interval = Math.max(1, recurring.rule.interval);
    const next = this.findNextOccurrence(recurring, startDate, fromDate, interval);
    if (!next) return null;
    if (next.getTime() > endAtMs) return null;
    return next.toISOString();
  }

  private computeDefaultRecurringAnchorDate(
    recurring: RecurringTriggerConfig,
    nowDate: Date,
  ): Date {
    const anchor = new Date(nowDate);
    anchor.setSeconds(0, 0);

    if (recurring.rule.type === 'hourly') {
      anchor.setMinutes(0, 0, 0);
      return anchor;
    }

    if (recurring.rule.type === 'daily') {
      const hour = recurring.rule.time_of_day?.hour ?? nowDate.getHours();
      const minute = recurring.rule.time_of_day?.minute ?? nowDate.getMinutes();
      anchor.setHours(hour, minute, 0, 0);
      return anchor;
    }

    const hour = recurring.rule.time_of_day?.hour ?? nowDate.getHours();
    const minute = recurring.rule.time_of_day?.minute ?? nowDate.getMinutes();
    const weekAnchor = startOfWeek(anchor);
    weekAnchor.setHours(hour, minute, 0, 0);
    return weekAnchor;
  }

  private findNextOccurrence(
    recurring: RecurringTriggerConfig,
    startDate: Date,
    fromDate: Date,
    interval: number,
  ): Date | null {
    const rule = recurring.rule;

    if (rule.type === 'hourly') {
      const baseMs = startDate.getTime();
      const fromMs = Math.max(fromDate.getTime(), baseMs);
      const intervalMs = interval * 60 * 60 * 1000;
      const steps = Math.ceil((fromMs - baseMs) / intervalMs);
      return new Date(baseMs + Math.max(0, steps) * intervalMs);
    }

    if (rule.type === 'daily') {
      const hour = rule.time_of_day?.hour ?? startDate.getHours();
      const minute = rule.time_of_day?.minute ?? startDate.getMinutes();
      const anchor = withTime(startDate, hour, minute);
      let candidate = new Date(anchor);
      if (candidate < fromDate) {
        const dayMs = 24 * 60 * 60 * 1000;
        const deltaDays = Math.floor((fromDate.getTime() - candidate.getTime()) / dayMs);
        const steps = Math.floor(deltaDays / interval);
        candidate = new Date(candidate.getTime() + steps * interval * dayMs);
        while (candidate < fromDate) {
          candidate = new Date(candidate.getTime() + interval * dayMs);
        }
      }
      return candidate;
    }

    if (rule.type === 'weekly') {
      const days = rule.weekly_rule?.days_of_week ?? [];
      if (days.length === 0) return null;
      const allowed = [...new Set(days.map((day) => DAY_INDEX[day]))].sort(
        (a, b) => a - b,
      );
      const hour = rule.time_of_day?.hour ?? startDate.getHours();
      const minute = rule.time_of_day?.minute ?? startDate.getMinutes();
      const weekMs = 7 * 24 * 60 * 60 * 1000;
      const anchorWeekStartMs = startOfWeek(startDate).getTime();
      const fromWeekStartMs = startOfWeek(fromDate).getTime();
      const rawWeekDiff = Math.floor((fromWeekStartMs - anchorWeekStartMs) / weekMs);
      const baselineWeekDiff = Math.max(0, rawWeekDiff);
      const remainder = baselineWeekDiff % interval;
      let alignedWeekDiff =
        remainder === 0
          ? baselineWeekDiff
          : baselineWeekDiff + (interval - remainder);

      for (let attempt = 0; attempt < 2; attempt += 1) {
        const weekStart = new Date(anchorWeekStartMs + alignedWeekDiff * weekMs);
        let earliestCandidate: Date | null = null;

        for (const dayIndex of allowed) {
          const candidate = new Date(weekStart);
          candidate.setDate(weekStart.getDate() + dayIndex);
          candidate.setHours(hour, minute, 0, 0);

          if (candidate < startDate || candidate < fromDate) continue;
          if (!earliestCandidate || candidate < earliestCandidate) {
            earliestCandidate = candidate;
          }
        }

        if (earliestCandidate) return earliestCandidate;
        alignedWeekDiff += interval;
      }

      return null;
    }

    return null;
  }

  private hasPendingForCampaignAt(
    snapshot: CampaignStateSnapshot,
    campaignId: string,
    executeAt: string,
  ): boolean {
    for (const pending of snapshot.queued_messages) {
      if (pending.campaign_id === campaignId && pending.execute_at === executeAt) {
        return true;
      }
    }
    return false;
  }

  private hasFuturePendingForCampaign(
    snapshot: CampaignStateSnapshot,
    campaignId: string,
    now: string,
  ): boolean {
    const nowMs = new Date(now).getTime();
    for (const pending of snapshot.queued_messages) {
      if (pending.campaign_id !== campaignId) continue;
      const executeAtMs = new Date(pending.execute_at).getTime();
      if (Number.isNaN(executeAtMs)) return true;
      if (!Number.isNaN(nowMs) && executeAtMs >= nowMs) return true;
    }
    return false;
  }

  private getCampaignState(
    snapshot: CampaignStateSnapshot,
    campaignId: string,
  ): CampaignStateRecord | undefined {
    return snapshot.campaign_states.find((state) => state.campaign_id === campaignId);
  }
}

File: skills/openclix-init/templates/android/engine/CampaignProcessor.kt
package ai.openclix.engine

import ai.openclix.models.Campaign
import ai.openclix.models.CampaignStateRecord
import ai.openclix.models.CampaignStateSnapshot
import ai.openclix.models.DayOfWeek
import ai.openclix.models.DecisionTrace
import ai.openclix.models.QueuedMessage
import ai.openclix.models.QueuedMessageContent
import ai.openclix.models.QueuedMessageStatus
import ai.openclix.models.RecurringTriggerConfig
import ai.openclix.models.Settings
import ai.openclix.models.SkipReason
import ai.openclix.models.TriggerContext
import ai.openclix.models.TriggerType
import ai.openclix.models.ClixLogger
import ai.openclix.services.renderTemplate
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

data class CampaignDecision(
    val trace: DecisionTrace,
    val action: String,
    val queued_message: QueuedMessage? = null,
    val scheduled_for: String? = null
)

data class CampaignProcessorDependencies(
    val eventConditionProcessor: EventConditionProcessor,
    val scheduleCalculator: ScheduleCalculator,
    val logger: ClixLogger,
    val settings: Settings? = null
)

private data class ExecutionResolution(
    val execute_at: String? = null,
    val scheduled_for: String? = null,
    val trigger_event_id: String? = null,
    val reason: String? = null,
    val skip_reason: SkipReason? = null
)

private fun createTrace(
    campaign_id: String,
    action: String,
    result: String,
    reason: String,
    skip_reason: SkipReason? = null
): DecisionTrace = DecisionTrace(
    campaign_id = campaign_id,
    action = action,
    result = result,
    skip_reason = skip_reason,
    reason = reason
)

private fun createSkipDecision(
    campaignId: String,
    reason: String,
    skipReason: SkipReason? = null
): CampaignDecision {
    return CampaignDecision(
        action = "skip",
        trace = createTrace(
            campaign_id = campaignId,
            action = "skip_campaign",
            result = "skipped",
            reason = reason,
            skip_reason = skipReason
        )
    )
}

private fun parseIso8601OrNull(isoString: String?): Long? {
    if (isoString.isNullOrBlank()) return null

    val formats = listOf(
        "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
        "yyyy-MM-dd'T'HH:mm:ss'Z'"
    )

    for (pattern in formats) {
        try {
            val formatter = SimpleDateFormat(pattern, Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
                isLenient = false
            }
            val parsedDate = formatter.parse(isoString)
            if (parsedDate != null) {
                return parsedDate.time
            }
        } catch (_: Exception) {
            continue
        }
    }

    return null
}

private fun startOfWeek(date: Date): Date {
    val calendar = Calendar.getInstance().apply {
        time = date
        set(Calendar.HOUR_OF_DAY, 0)
        set(Calendar.MINUTE, 0)
        set(Calendar.SECOND, 0)
        set(Calendar.MILLISECOND, 0)
        add(Calendar.DAY_OF_MONTH, -get(Calendar.DAY_OF_WEEK) + Calendar.SUNDAY)
    }
    return calendar.time
}

private fun withTime(baseDate: Date, hour: Int, minute: Int): Date {
    val calendar = Calendar.getInstance().apply {
        time = baseDate
        set(Calendar.HOUR_OF_DAY, hour)
        set(Calendar.MINUTE, minute)
        set(Calendar.SECOND, 0)
        set(Calendar.MILLISECOND, 0)
    }
    return calendar.time
}

class CampaignProcessor {

    fun process(
        campaignId: String,
        campaign: Campaign,
        context: TriggerContext,
        snapshot: CampaignStateSnapshot,
        dependencies: CampaignProcessorDependencies
    ): CampaignDecision {
        val eventConditionProcessor = dependencies.eventConditionProcessor
        val scheduleCalculator = dependencies.scheduleCalculator
        val logger = dependencies.logger
        val settings = dependencies.settings

        val now = context.now ?: toIso8601(System.currentTimeMillis())
        val campaignState = getCampaignState(snapshot, campaignId)

        if (campaign.status.value != "running") {
            val reason = "Campaign status is '${campaign.status.value}', not 'running'"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(campaignId, reason, SkipReason.CAMPAIGN_NOT_RUNNING)
        }

        if (
            (campaign.trigger.type == TriggerType.EVENT && context.trigger != "event_tracked") ||
            (campaign.trigger.type != TriggerType.EVENT && context.trigger == "event_tracked")
        ) {
            val reason = "Trigger type '${campaign.trigger.type.value}' is not eligible for '${context.trigger}'"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(campaignId, reason)
        }

        if (campaign.trigger.type != TriggerType.RECURRING && campaignState?.triggered == true) {
            val reason = "Campaign already triggered"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(campaignId, reason)
        }

        if (settings?.frequency_cap != null) {
            val frequencyCap = settings.frequency_cap
            val windowStartEpoch = parseIso8601(now) - frequencyCap.window_seconds * 1000L
            val windowStart = toIso8601(windowStartEpoch)
            val countInWindow = snapshot.trigger_history.count { historyRow ->
                historyRow.triggered_at >= windowStart
            }

            if (countInWindow >= frequencyCap.max_count) {
                val reason = "Frequency cap exceeded ($countInWindow/${frequencyCap.max_count} within ${frequencyCap.window_seconds}s)"
                logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
                return createSkipDecision(
                    campaignId,
                    reason,
                    SkipReason.CAMPAIGN_FREQUENCY_CAP_EXCEEDED
                )
            }
        }

        if (campaign.trigger.type == TriggerType.RECURRING && hasFuturePendingForCampaign(snapshot, campaignId, now)) {
            val reason = "Recurring campaign already has a queued message"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(campaignId, reason)
        }

        val executionResolution = resolveExecutionTime(
            campaign = campaign,
            context = context,
            campaignState = campaignState,
            now = now,
            eventConditionProcessor = eventConditionProcessor
        )

        if (executionResolution.execute_at == null) {
            val reason = executionResolution.reason ?: "Campaign trigger conditions were not met"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(campaignId, reason, executionResolution.skip_reason)
        }

        if (hasPendingForCampaignAt(snapshot, campaignId, executionResolution.execute_at)) {
            val reason = "Duplicate schedule prevented for campaign at ${executionResolution.execute_at}"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(campaignId, reason)
        }

        val scheduleResult = scheduleCalculator.calculate(
            ScheduleInput(
                now = now,
                execute_at = executionResolution.execute_at,
                do_not_disturb = settings?.do_not_disturb
            )
        )

        if (scheduleResult.skipped) {
            val reason = "Blocked by do-not-disturb window"
            logger.debug("[CampaignProcessor] Campaign $campaignId skipped: $reason")
            return createSkipDecision(
                campaignId,
                reason,
                scheduleResult.skip_reason
            )
        }

        val templateVariables = mutableMapOf<String, Any?>()
        context.event?.properties?.let { eventProperties ->
            templateVariables.putAll(eventProperties)
        }

        val renderedTitle = renderTemplate(campaign.message.content.title, templateVariables)
        val renderedBody = renderTemplate(campaign.message.content.body, templateVariables)

        val queuedMessage = QueuedMessage(
            id = UUID.randomUUID().toString(),
            campaign_id = campaignId,
            channel_type = campaign.message.channel_type,
            status = QueuedMessageStatus.SCHEDULED,
            content = QueuedMessageContent(
                title = renderedTitle,
                body = renderedBody,
                image_url = campaign.message.content.image_url,
                landing_url = campaign.message.content.landing_url
            ),
            trigger_event_id = executionResolution.trigger_event_id,
            execute_at = scheduleResult.execute_at,
            created_at = now
        )

        logger.debug(
            "[CampaignProcessor] Campaign $campaignId: triggered, scheduled for ${scheduleResult.execute_at}"
        )

        return CampaignDecision(
            action = "trigger",
            trace = createTrace(
                campaign_id = campaignId,
                action = "trigger_campaign",
                result = "applied",
                reason = "Campaign triggered, message scheduled for ${scheduleResult.execute_at}"
            ),
            queued_message = queuedMessage,
            scheduled_for = executionResolution.scheduled_for ?: scheduleResult.execute_at
        )
    }

    private fun resolveExecutionTime(
        campaign: Campaign,
        context: TriggerContext,
        campaignState: CampaignStateRecord?,
        now: String,
        eventConditionProcessor: EventConditionProcessor
    ): ExecutionResolution {
        if (campaign.trigger.type == TriggerType.EVENT) {
            val eventConfig = campaign.trigger.event
            if (eventConfig == null) {
                return ExecutionResolution(
                    reason = "Trigger type 'event' requires trigger.event configuration",
                    skip_reason = SkipReason.TRIGGER_EVENT_NOT_MATCHED
                )
            }

            if (context.event == null) {
                return ExecutionResolution(
                    reason = "Event trigger requires an event in context",
                    skip_reason = SkipReason.TRIGGER_EVENT_NOT_MATCHED
                )
            }

            val isMatched = eventConditionProcessor.process(eventConfig.trigger_event, context.event)
            if (!isMatched) {
                return ExecutionResolution(
                    reason = "Trigger event conditions did not match event '${context.event.name}'",
                    skip_reason = SkipReason.TRIGGER_EVENT_NOT_MATCHED
                )
            }

            val delaySeconds = eventConfig.delay_seconds ?: 0
            val executeAt = toIso8601(parseIso8601(now) + delaySeconds * 1000L)
            return ExecutionResolution(
                execute_at = executeAt,
                scheduled_for = executeAt,
                trigger_event_id = context.event.id
            )
        }

        if (campaign.trigger.type == TriggerType.SCHEDULED) {
            val scheduledConfig = campaign.trigger.scheduled
            if (scheduledConfig == null || parseIso8601OrNull(scheduledConfig.execute_at) == null) {
                return ExecutionResolution(
                    reason = "Scheduled trigger requires a valid execute_at datetime"
                )
            }

            val executeAtEpoch = parseIso8601OrNull(scheduledConfig.execute_at) ?: return ExecutionResolution(
                reason = "Scheduled trigger requires a valid execute_at datetime"
            )
            val nowEpoch = parseIso8601(now)
            if (executeAtEpoch <= nowEpoch) {
                return ExecutionResolution(
                    reason = "Scheduled execute_at '${scheduledConfig.execute_at}' is already in the past"
                )
            }

            return ExecutionResolution(
                execute_at = scheduledConfig.execute_at,
                scheduled_for = scheduledConfig.execute_at
            )
        }

        val recurringConfig = campaign.trigger.recurring
        if (recurringConfig == null) {
            return ExecutionResolution(reason = "Recurring trigger requires recurring configuration")
        }

        val nextExecuteAt = computeNextRecurringExecuteAt(
            recurring = recurringConfig,
            now = now,
            lastScheduledAt = campaignState?.recurring_last_scheduled_at,
            recurringAnchorAt = campaignState?.recurring_anchor_at
        )

        if (nextExecuteAt == null) {
            return ExecutionResolution(reason = "Recurring schedule has no upcoming execution window")
        }

        return ExecutionResolution(
            execute_at = nextExecuteAt,
            scheduled_for = nextExecuteAt
        )
    }

    private fun computeNextRecurringExecuteAt(
        recurring: RecurringTriggerConfig,
        now: String,
        lastScheduledAt: String?,
        recurringAnchorAt: String?
    ): String? {
        val nowEpoch = parseIso8601OrNull(now) ?: return null
        val nowDate = Date(nowEpoch)

        val startDate = when {
            parseIso8601OrNull(recurring.start_at) != null -> Date(parseIso8601OrNull(recurring.start_at)!!)
            parseIso8601OrNull(recurringAnchorAt) != null -> Date(parseIso8601OrNull(recurringAnchorAt)!!)
            else -> computeDefaultRecurringAnchorDate(recurring, nowDate)
        }

        val endEpoch = if (recurring.end_at.isNullOrBlank()) {
            Long.MAX_VALUE
        } else {
            parseIso8601OrNull(recurring.end_at) ?: return null
        }

        val fromDate = if (parseIso8601OrNull(lastScheduledAt) != null) {
            Date(parseIso8601OrNull(lastScheduledAt)!! + 60_000L)
        } else {
            nowDate
        }

        val interval = maxOf(1, recurring.rule.interval)
        val nextOccurrence = findNextOccurrence(recurring, startDate, fromDate, interval) ?: return null
        if (nextOccurrence.time > endEpoch) return null

        return toIso8601(nextOccurrence.time)
    }

    private fun computeDefaultRecurringAnchorDate(recurring: RecurringTriggerConfig, nowDate: Date): Date {
        val anchorCalendar = Calendar.getInstance().apply {
            time = nowDate
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }

        if (recurring.rule.type.value == "hourly") {
            anchorCalendar.set(Calendar.MINUTE, 0)
            return anchorCalendar.time
        }

        if (recurring.rule.type.value == "daily") {
            val hour = recurring.rule.time_of_day?.hour ?: anchorCalendar.get(Calendar.HOUR_OF_DAY)
            val minute = recurring.rule.time_of_day?.minute ?: anchorCalendar.get(Calendar.MINUTE)
            anchorCalendar.set(Calendar.HOUR_OF_DAY, hour)
            anchorCalendar.set(Calendar.MINUTE, minute)
            return anchorCalendar.time
        }

        val hour = recurring.rule.time_of_day?.hour ?: anchorCalendar.get(Calendar.HOUR_OF_DAY)
        val minute = recurring.rule.time_of_day?.minute ?: anchorCalendar.get(Calendar.MINUTE)
        val weekStartDate = startOfWeek(anchorCalendar.time)
        return withTime(weekStartDate, hour, minute)
    }

    private fun findNextOccurrence(
        recurring: RecurringTriggerConfig,
        startDate: Date,
        fromDate: Date,
        interval: Int
    ): Date? {
        val recurrenceRule = recurring.rule

        if (recurrenceRule.type.value == "hourly") {
            val baseMilliseconds = startDate.time
            val fromMilliseconds = maxOf(fromDate.time, baseMilliseconds)
            val intervalMilliseconds = interval * 60L * 60L * 1000L
            val steps = if (fromMilliseconds <= baseMilliseconds) {
                0L
            } else {
                kotlin.math.ceil((fromMilliseconds - baseMilliseconds).toDouble() / intervalMilliseconds.toDouble()).toLong()
            }
            return Date(baseMilliseconds + steps * intervalMilliseconds)
        }

        if (recurrenceRule.type.value == "daily") {
            val hour = recurrenceRule.time_of_day?.hour ?: Calendar.getInstance().apply { time = startDate }.get(Calendar.HOUR_OF_DAY)
            val minute = recurrenceRule.time_of_day?.minute ?: Calendar.getInstance().apply { time = startDate }.get(Calendar.MINUTE)
            var candidate = withTime(startDate, hour, minute)
            if (candidate.before(fromDate)) {
                val dayMilliseconds = 24L * 60L * 60L * 1000L
                val deltaDays = ((fromDate.time - candidate.time) / dayMilliseconds).toInt()
                val steps = deltaDays / interval
                candidate = Date(candidate.time + steps * interval * dayMilliseconds)
                while (candidate.before(fromDate)) {
                    candidate = Date(candidate.time + interval * dayMilliseconds)
                }
            }
            return candidate
        }

        if (recurrenceRule.type.value == "weekly") {
            val daysOfWeek = recurrenceRule.weekly_rule?.days_of_week ?: emptyList()
            if (daysOfWeek.isEmpty()) return null

            val allowedDays = daysOfWeek
                .map(DayOfWeek::index)
                .distinct()
                .sorted()
            val startCalendar = Calendar.getInstance().apply { time = startDate }
            val hour = recurrenceRule.time_of_day?.hour ?: startCalendar.get(Calendar.HOUR_OF_DAY)
            val minute = recurrenceRule.time_of_day?.minute ?: startCalendar.get(Calendar.MINUTE)

            val weekMilliseconds = 7L * 24L * 60L * 60L * 1000L
            val anchorWeekStartMilliseconds = startOfWeek(startDate).time
            val fromWeekStartMilliseconds = startOfWeek(fromDate).time
            val rawWeekDiff = kotlin.math.floor(
                (fromWeekStartMilliseconds - anchorWeekStartMilliseconds).toDouble() / weekMilliseconds.toDouble()
            ).toInt()
            val baselineWeekDiff = maxOf(0, rawWeekDiff)
            val remainder = baselineWeekDiff % interval
            var alignedWeekDiff = if (remainder == 0) {
                baselineWeekDiff
            } else {
                baselineWeekDiff + (interval - remainder)
            }

            repeat(2) {
                val weekStartDate = Date(anchorWeekStartMilliseconds + alignedWeekDiff * weekMilliseconds)
                var earliestCandidate: Date? = null

                for (dayIndex in allowedDays) {
                    val candidateCalendar = Calendar.getInstance().apply {
                        time = weekStartDate
                        add(Calendar.DAY_OF_MONTH, dayIndex)
                        set(Calendar.HOUR_OF_DAY, hour)
                        set(Calendar.MINUTE, minute)
                        set(Calendar.SECOND, 0)
                        set(Calendar.MILLISECOND, 0)
                    }
                    val candidate = candidateCalendar.time

                    if (candidate.before(startDate) || candidate.before(fromDate)) continue
                    if (earliestCandidate == null || candidate.before(earliestCandidate)) {
                        earliestCandidate = candidate
                    }
                }

                if (earliestCandidate != null) {
                    return earliestCandidate
                }

                alignedWeekDiff += interval
            }

            return null
        }

        return null
    }

    private fun hasPendingForCampaignAt(
        snapshot: CampaignStateSnapshot,
        campaignId: String,
        executeAt: String
    ): Boolean {
        return snapshot.queued_messages.any { queuedMessage ->
            queuedMessage.campaign_id == campaignId && queuedMessage.execute_at == executeAt
        }
    }

    private fun hasFuturePendingForCampaign(
        snapshot: CampaignStateSnapshot,
        campaignId: String,
        now: String
    ): Boolean {
        val nowEpoch = parseIso8601OrNull(now)
        for (pendingMessage in snapshot.queued_messages) {
            if (pendingMessage.campaign_id != campaignId) continue

            val executeAtEpoch = parseIso8601OrNull(pendingMessage.execute_at)
            if (executeAtEpoch == null) return true
            if (nowEpoch != null && executeAtEpoch >= nowEpoch) return true
        }
        return false
    }

    private fun getCampaignState(
        snapshot: CampaignStateSnapshot,
        campaignId: String
    ): CampaignStateRecord? {
        return snapshot.campaign_states.firstOrNull { row -> row.campaign_id == campaignId }
    }
}
